rust/src/mqtt/mqtt_message.rs

```diff
@@ -0,0 +1,187 @@
+    pub username_flag: bool,
+    pub password_flag: bool,
+    pub will_retain: bool,
+    pub will_qos: u8,
+    pub will_flag: bool,
+    pub clean_session: bool,
+    pub keepalive: u16,
+    pub client_id: String,
+    pub will_topic: Option<String>,
+    pub will_message: Option<Vec<u8>>,
+    pub username: Option<String>,
+    pub password: Option<Vec<u8>>,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+    pub will_properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTConnackData {
+    pub return_code: u8,
+    pub session_present: bool,                 // MQTT 3.1.1
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTPublishData {
+    pub topic: String,
+    pub message_id: Option<u16>,
+    pub message: Vec<u8>,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTMessageIdOnly {
+    pub message_id: u16,
+    pub reason_code: Option<u8>,               // MQTT 5.0
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTSubscribeTopicData {
+    pub topic_name: String,
+    pub qos: u8,
+}
+
+#[derive(Debug)]
+pub struct MQTTSubscribeData {
+    pub message_id: u16,
+    pub topics: Vec<MQTTSubscribeTopicData>,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTSubackData {
+    pub message_id: u16,
+    pub qoss: Vec<u8>,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTUnsubscribeData {
+    pub message_id: u16,
+    pub topics: Vec<String>,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTUnsubackData {
+    pub message_id: u16,
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+    pub reason_codes: Option<Vec<u8>>,         // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTAuthData {
+    pub reason_code: u8,                       // MQTT 5.0
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTDisconnectData {
+    pub reason_code: Option<u8>,               // MQTT 5.0
+    pub properties: Option<Vec<MQTTProperty>>, // MQTT 5.0
+}
+
+#[derive(Debug)]
+pub struct MQTTTruncatedData {
+    pub original_message_type: u8,
+    pub skipped_length: usize,
+}
+
+pub fn message_code_string(op: u8) -> String {
+    match op {
+        1 => "connect",
+        2 => "connack",
+        3 => "publish",
+        4 => "puback",
+        5 => "pubrec",
+        6 => "pubrel",
+        7 => "pubcomp",
+        8 => "subscribe",
+        9 => "suback",
+        10 => "unsubscribe",
+        11 => "unsuback",
+        12 => "pingreq",
+        13 => "pingresp",
+        14 => "disconnect",
+        15 => "auth",
+        _ => "unassigned",
+    }
+    .to_string()
+}
+
+#[no_mangle]
+pub extern "C" fn rs_cstr_message_code(
+    op: *mut u8,
```

Comment:
I wonder if we should cheat and have this in the return value...

Reply:
see also my idea for `rs_http2_parse_frametype` where I used `from_str` to be more inline with Rust usual naming
https://github.com/OISF/suricata/pull/5179/files#diff-18b926fc6ceba8f9d968d8884cc4e9f3R53
https://github.com/OISF/suricata/pull/5179/files#diff-1731300ec3665501b1d0927c8ae1dc40R49

Reply:
> I wonder if we should cheat and have this in the return value...

OK, entirely possible. Can change this.

> see also my idea for rs_http2_parse_frametype where I used from_str to be more inline with Rust usual naming

Well, your enum for `HTTP2FrameType` is simply numeric and can be used as an integer -- mine carry associated values. I was under the impression that these cannot be represented as integers, as that would need `FromPrimitive` which I can't use here:
```
error: proc-macro derive panicked
  --> src/mqtt/mqtt_message.rs:31:10
   |
31 | #[derive(FromPrimitive, Debug)]
   |          ^^^^^^^^^^^^^
   |
   = help: message: `FromPrimitive` can be applied only to unitary enums and newtypes, MQTTOperation::CONNECT is either struct or tuple
```
So to me it looks like I really need an independent mapping between the string representation (e.g. `connect`, which can not be derived from the `MQTTOperation` enum, but needed e.g. for printing the sub-items in the EVE-JSON logger) and the numeric codes (which are used to match against the parsed header field, but cannot be derived from the enum either).
Did I miss something?

Reply:
Yes the MQTT `message_type` is `u8` and not a specific enum so we cannot use `FromPrimitive` 

So, you can either create a specific enum for `message_type`

Or maybe just use a `match` instead of a `for` loop

Reply:
I still do not see how that wold help me convert a string to an integer (even if it is an enum) and vice versa. I would still need to either define the mapping bidirectionally:

- mapping the string to the number so I can use `match` on the string to get the numeric code from the signature keyword
- mapping the number  to the string so I can use `match`on the number to get the string for JSON logging

I wanted to avoid that, so I used a loop.
Could you explain what you mean? I seem to be a bit lost seeing the improvement here.

Reply:
Ok so it is a case of the double mapping versus the `loop`

If you prefer avoid the double mapping, it is ok for me (you can add a comment about it)

Reply:
I have reworked the handling of these values to use a specific enum. See next PR.

Reply:
Ok also :-)